heimlig/src/hsm/workers/chachapoly_worker.rs

```diff
@@ -1,83 +1,70 @@
-        aad: Option<PoolChunk>,
-        mut ciphertext: PoolChunk,
-    ) -> Response {
-        match self.pool.alloc(crate::crypto::chacha20poly1305::TAG_SIZE) {
-            Err(_) => Response::Error(Error::Alloc),
-            Ok(mut tag) => {
-                let aad = match &aad {
-                    Some(aad) => aad.as_slice(),
-                    None => &[] as &[u8],
-                };
-                match crate::crypto::chacha20poly1305::encrypt_in_place_detached(
-                    key,
-                    nonce.as_slice(),
-                    aad,
-                    ciphertext.as_slice_mut(),
-                ) {
-                    Ok(computed_tag) => {
-                        if computed_tag.len() != tag.as_slice().len() {
-                            return Response::Error(Error::Alloc);
-                        }
-                        tag.as_slice_mut().copy_from_slice(computed_tag.as_slice());
-                        Response::EncryptChaChaPoly { ciphertext, tag }
-                    }
-                    Err(e) => Response::Error(Error::Crypto(e)),
+        nonce: &[u8],
+        aad: Option<&[u8]>,
+        ciphertext: &'a mut [u8],
+        tag: &'a mut [u8],
+    ) -> Response<'a> {
+        let aad = match &aad {
+            Some(aad) => aad,
+            None => &[] as &[u8],
+        };
+        match crate::crypto::chacha20poly1305::encrypt_in_place_detached(
+            key, nonce, aad, ciphertext,
+        ) {
+            Ok(computed_tag) => {
+                if computed_tag.len() != tag.len() {
+                    return Response::Error(Error::Alloc);
                 }
+                tag.copy_from_slice(computed_tag.as_slice());
+                Response::EncryptChaChaPoly { ciphertext, tag }
             }
+            Err(e) => Response::Error(Error::Crypto(e)),
         }
     }
 
-    pub fn decrypt_external_key(
+    pub fn decrypt_external_key<'a>(
         &mut self,
-        key: PoolChunk,
-        nonce: PoolChunk,
-        aad: Option<PoolChunk>,
-        plaintext: PoolChunk,
-        tag: PoolChunk,
-    ) -> Response {
-        self.decrypt(key.as_slice(), nonce, aad, plaintext, tag)
+        key: &[u8],
+        nonce: &[u8],
+        aad: Option<&[u8]>,
+        plaintext: &'a mut [u8],
+        tag: &[u8],
+    ) -> Response<'a> {
+        self.decrypt(key, nonce, aad, plaintext, tag)
     }
 
-    pub fn decrypt(
+    pub fn decrypt<'a>(
         &mut self,
         key: &[u8],
-        nonce: PoolChunk,
-        aad: Option<PoolChunk>,
-        mut plaintext: PoolChunk,
-        tag: PoolChunk,
-    ) -> Response {
+        nonce: &[u8],
+        aad: Option<&[u8]>,
+        plaintext: &'a mut [u8],
+        tag: &[u8],
+    ) -> Response<'a> {
         let aad = match &aad {
-            Some(aad) => aad.as_slice(),
+            Some(aad) => aad,
             None => &[] as &[u8],
         };
         match crate::crypto::chacha20poly1305::decrypt_in_place_detached(
```

Comment:
shall we check tag size before calling decrypt_in_place_detached? or does it do a check internally? if so, what is the correct size?

Reply:
`decrypt_in_place_detached` checks the tag size.
https://github.com/eclipse-heimlig/heimlig/blob/cd3b0689f5e5a339b6aa1e3cd6ec37898a90d10a/heimlig/src/crypto/chacha20poly1305.rs#L60

The `TAG_SIZE` is 16 bytes.
https://github.com/eclipse-heimlig/heimlig/blob/cd3b0689f5e5a339b6aa1e3cd6ec37898a90d10a/heimlig/src/crypto/chacha20poly1305.rs#L12

Reply:
See proposed change in [f970059](https://github.com/eclipse-heimlig/heimlig/pull/157/commits/f9700595fa9ef8af1c5f43bfe401074a38b75170), i.e., make it explicit in the types that the tag should have the correct size; this means offloading the error handling to the user; see for instance [the test in scheduler](https://github.com/eclipse-heimlig/heimlig/pull/157/commits/f9700595fa9ef8af1c5f43bfe401074a38b75170#diff-bb6a5fb883712e7e170d36fd6dd4d53701e97fdd7a20acd6deea484098c8f452R108).